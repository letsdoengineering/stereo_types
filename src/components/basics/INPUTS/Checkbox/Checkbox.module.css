/* Customize the label (the container) */
.container {
    display: flex;
    align-items: center;
    position: relative;
    margin-left: 16px;
    cursor: pointer;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

/* Hide the browser's default checkbox */
.container input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
}

/* Add style for when focus on original input style to our checkmark */
.container input:focus ~ .checkmark {
    -moz-box-shadow:    0 0 0 2px var(--focus);
    -webkit-box-shadow: 0 0 0 2px var(--focus);
    box-shadow:         0 0 0 2px var(--focus);
}

.label {
    font-size: var(--font-size-S);
    color: var(--dark-gray);
}
/* Create a custom checkbox */
.checkmark {
    margin-right: 4px;
    height: 1rem;
    width: 1rem;
    border: 2px solid var(--dark-gray);
    background-color: white;
}

/* When checked add background */
.container input:checked ~ .checkmark {
    background-color: var(--dark-gray);
}

/* When checked but unavailable add gray background */
.container input:checked ~ .checkmark.unavailable {
    background-color: var(--outline-gray);
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark:after {
    content: "";
    position: absolute;
    display: none;
}

/* Show the checkmark when checked */
.container input:checked ~ .checkmark:after {
    display: block;
}

/* Style the checkmark/indicator */
.container .checkmark:after {
    left: 5px;
    width: 6px;
    height: 10px;
    border: solid white;
    border-width: 0 3px 3px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
}

.unavailable {
    color: var(--outline-gray);
    border-color: var(--outline-gray);
    cursor: default;
}
