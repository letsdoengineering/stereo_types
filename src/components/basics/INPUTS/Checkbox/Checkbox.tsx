import React, { FunctionComponent } from 'react'
import Text from '../../Text/Text'
import * as checkboxStyles from './Checkbox.module.css'
import getClassNames from '../../../../utils/get-class-names'

export type CheckboxProps = React.ComponentProps<'input'> &
    Readonly<{
        /** id used for id of input - custom styling */
        id?: string | undefined
        /** tell the checkbox what it's status is, controlled by consumer and triggered via onChange */
        isChecked: boolean
        /** Called when toggled */
        onChange(): void
        /** value you wish the checkbox to have */
        value: string
        /** text for showing next to checkbox */
        text: string
        /** toggle to make unavailable */
        available?: boolean
    }>

const Checkbox: FunctionComponent<CheckboxProps> = ({
    isChecked,
    id,
    available = true,
    onChange,
    value,
    text,
    ...rest
}) => {
    const textClassName = getClassNames({
        conditionalClasses: { [`${checkboxStyles.unavailable}`]: !available },
    })
    const checkboxClassName = getClassNames({
        defaultClasses: [`${checkboxStyles.checkmark}`],
        conditionalClasses: { [`${checkboxStyles.unavailable}`]: !available },
    })

    return (
        <label className={checkboxStyles.container}>
            <input
                disabled={!available}
                checked={isChecked}
                id={id}
                onChange={onChange}
                type='checkbox'
                value={value}
                {...rest}
            />
            <span className={checkboxClassName} />
            <Text className={textClassName}>{text}</Text>
        </label>
    )
}

export default Checkbox
