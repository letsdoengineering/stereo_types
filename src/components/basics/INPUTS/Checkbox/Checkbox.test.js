import React from 'react'
import Checkbox from './Checkbox'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { axe } from 'jest-axe'

describe('Checkbox', () => {
  it('should call onChange when clicked', () => {
    const onChange = jest.fn()
    render(<Checkbox isChecked={true} onChange={onChange} value={'text shown as label'} />)
    const checkbox = screen.getByLabelText('text shown as label')
    userEvent.click(checkbox)
    expect(onChange).toHaveBeenCalled()
  })

  it('should not call onChange when clicked if available is false', () => {
    const onChange = jest.fn()
    render(
      <Checkbox
        isChecked={true}
        available={false}
        onChange={onChange}
        value={'text shown as label'}
      />
    )
    const checkbox = screen.getByLabelText('text shown as label')
    userEvent.click(checkbox)
    expect(onChange).not.toHaveBeenCalled()
  })

  it('should render correctly', () => {
    const onChange = jest.fn()
    const tree = renderer
      .create(<Checkbox isChecked={false} onChange={onChange} value={'text shown as label'} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders the page DOM that is accessible', async () => {
    const onChange = jest.fn()
    const { container } = render(
      <Checkbox isChecked={true} onChange={onChange} value={'text shown as label'} />
    )
    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
