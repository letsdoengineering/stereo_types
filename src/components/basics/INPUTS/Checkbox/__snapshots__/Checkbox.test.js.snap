// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
<label
  className="container"
>
  <input
    checked={false}
    disabled={false}
    onChange={[MockFunction]}
    type="checkbox"
    value="text shown as label"
  />
  <span
    className="checkmark"
  />
  <span
    className="sizeS weightNormal textGray"
  >
    text shown as label
  </span>
</label>
`;
